physical/etcd3.go

```diff
@@ -0,0 +1,285 @@
+package physical
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	metrics "github.com/armon/go-metrics"
+	"github.com/coreos/etcd/client"
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+	"github.com/coreos/etcd/pkg/transport"
+	log "github.com/mgutz/logxi/v1"
+	"golang.org/x/net/context"
+)
+
+// Etcd3Backend is a physical backend that stores data at specific
+// prefix within Etcd. It is used for most production situations as
+// it allows Vault to run on multiple machines in a highly-available manner.
+type Etcd3Backend struct {
+	logger    log.Logger
+	path      string
+	haEnabled bool
+
+	permitPool *PermitPool
+
+	etcd *clientv3.Client
+}
+
+// newEtcd3Backend constructs a etcd3 backend.
+func newEtcd3Backend(conf map[string]string, logger log.Logger) (Backend, error) {
+	// Get the etcd path form the configuration.
+	path, ok := conf["path"]
+	if !ok {
+		path = "/vault"
+	}
+
+	// Ensure path is prefixed.
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+
+	// Set a default machines list and check for an overriding address value.
+	endpoints := []string{"http://127.0.0.1:2379"}
+	if address, ok := conf["address"]; ok {
+		endpoints = strings.Split(address, ",")
+	}
+
+	cfg := clientv3.Config{
+		Endpoints: endpoints,
+	}
+
+	haEnabled := os.Getenv("ETCD_HA_ENABLED")
+	if haEnabled == "" {
+		haEnabled = conf["ha_enabled"]
+	}
+	haEnabledBool, _ := strconv.ParseBool(haEnabled)
+
+	cert, hasCert := conf["tls_cert_file"]
+	key, hasKey := conf["tls_key_file"]
+	ca, hasCa := conf["tls_ca_file"]
+	if (hasCert && hasKey) || hasCa {
+		tls := transport.TLSInfo{
+			CAFile:   ca,
+			CertFile: cert,
+			KeyFile:  key,
+		}
+
+		tlscfg, err := tls.ClientConfig()
+		if err != nil {
+			return nil, err
+		}
+		cfg.TLS = tlscfg
+	}
+
+	// Set credentials.
+	username := os.Getenv("ETCD_USERNAME")
+	if username == "" {
+		username, _ = conf["username"]
+	}
+
+	password := os.Getenv("ETCD_PASSWORD")
+	if password == "" {
+		password, _ = conf["password"]
+	}
+
+	if username != "" && password != "" {
+		cfg.Username = username
+		cfg.Password = password
+	}
+
+	etcd, err := clientv3.New(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	sync, ok := conf["sync"]
+	if !ok {
+		sync = "yes"
+	}
+	switch sync {
+	case "yes", "true", "y", "1":
```

Comment:
Is the ability to use `yes` something etcd-specific? Otherwise I'd prefer `strconv.ParseBool` behavior to keep things consistent with most of the rest of the codebase.